### PR TITLE
Add getting the port out of a Server

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
For #72

### CODE REVIEW

File: core/src/main/java/io/grpc/Server.java [hunk 1]
```diff
@@ -52,6 +52,18 @@
   public abstract Server start() throws IOException;
 
   /**
+   * Returns the port number the server is listening on.  This can return -1 if there is no actual
+   * port or the result otherwise does not make sense.  Result is undefined after the server is
+   * terminated.
+   *
+   * @return the port number.
```
Comment (buchgr) [inline core/src/main/java/io/grpc/Server.java]: remove? imho doesn't seem useful after the description?

File: core/src/main/java/io/grpc/Server.java [hunk 2]
```diff
@@ -52,6 +52,18 @@
...
+   */
+  @ExperimentalApi
+  public int getPort() {
```
Comment (buchgr) [inline core/src/main/java/io/grpc/Server.java]: Why not make this an `Integer` and declare it `@Nullable`?

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/Server.java]: Because -1 is how URL reports an absent port number.  Precedent in the standard library, basically.

File: netty/src/test/java/io/grpc/netty/NettyServerTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.net.InetSocketAddress;
+
+@RunWith(JUnit4.class)
+public class NettyServerTest {
```
Comment (buchgr) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: why not simply use the `NettyServerBuilder` and `TestUtils.pickUnusedPort()`? I think I am missing something? :sweat_smile:

File: core/src/main/java/io/grpc/Server.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/Server.java]: Done.  Also added comment about calling before start.

File: netty/src/test/java/io/grpc/netty/NettyServerTest.java (hunk 3 above)
Comment (carl-mastrangelo) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: This checks that trying to open port zero actually opens some other ephemeral port.

Comment (buchgr) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: and why is this relevant to your changes? My understanding is that in `getPort()` you just want to test whether `NettyServer.getPort` returns the port it's actually listening on? Similar in `getPort_notStarted`.

Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: It seems to be testing the exact use case that @buchgr wanted in #72, right?

Comment (carl-mastrangelo) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: This is testing the expected use case: Listen on port 0, get some other port.  I want the call to go all the way down to the kernel and back.

pickUnusedPort is for testing?  Also it precludes you from picking the interface on which to pick an ephemeral port?

File: netty/src/test/java/io/grpc/netty/NettyServerTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    // Check that we got an actual port.
+    assertThat(ns.getPort()).isGreaterThan(0);
+
+    // Cleanup
+    ns.shutdown();
+  }
+
+  @Test
+  public void getPort_notStarted() throws Exception {
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: What is this test case testing? Isn't this testing an invalid state?

**Comment by ejona86** [comment]
@carl-mastrangelo LGTM

File: netty/src/test/java/io/grpc/netty/NettyServerTest.java (hunk 3 above)
Comment (buchgr) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: hmm, then maybe I found the test name misleading. In that case, why not do `NettyServerBuilder.getPort(0).build()` then?

Comment (buchgr) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: @carl-mastrangelo @ejona86 
To be clear, I meant :sweat_smile:
- Use `NettyServerBuilder` instead of directly calling `NettyServer`. Cause it's much shorter.
- Have one test that tests for port 0.
- Have one test that verifies that if the server binds on port X, then `getPort()` also returns X
- Have one test that tests not started.

No?

Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: @buchgr, that sounds fine to me; I've been using the build generally to create the `NettyServer`. Probably doesn't need a test for not started here, since it is invalid.

Comment (buchgr) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: > Probably doesn't need a test for not started here, since it is invalid.

@ejona86 yes you are right. I was mixing up `grpc.Server` and `grpc.internal.Server`, as the first one is supposed to throw a `IllegalStateException`, but that's tested in `ServerImplTest`.

Comment (carl-mastrangelo) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: NettyServerBuilder.build() doesn't return a NettyServer, surprisingly.  I don't want to involve too much of outside class functionality in this test, because it is out of scope.  (I.e. I don't want to have to call start on ServerImpl and have one of the side effects be that it calls NettyServer.start.)

This is a unit test, not an integration test, which is kind of what you are asking for: a test that goes through multiple parts of system.

File: netty/src/test/java/io/grpc/netty/NettyServerTest.java (hunk 4 above)
Comment (carl-mastrangelo) [inline netty/src/test/java/io/grpc/netty/NettyServerTest.java]: It's checking that the server returns -1 if there was a failure to start it, or not started.  It also bumps coverage up slightly, since it checks another code path.
